### PR TITLE
Part1

### PR DESCRIPTION
**Pull request by ffiorani**
We would like to merge our Part 1 additions to the main branch.

We implemented two dropboxes to let the user select their journey, and a button that redirects them to the corresponding LNER webpage. We added a little bit of styling.

Update: refactored code into a new StationDropDown file
image

### CODE REVIEW

File: src/components/Stations.tsx [hunk 1]
```diff
@@ -1,22 +1,73 @@
...
+    const [arrivalStationChoice, setArrivalStationChoices] = useState('PAD');
+
+    // useEffect(() => {
+    //  fetchStations()
+    // .then(response => response.json())
+    // .then(list => list.stations[0].id)
+    // .then((value) => alert(value))
+    // .catch((err) => alert(err))
+    // .finally(() => alert('finally'));
+    // }, []);
```
Comment (emma-bragg) [inline src/components/Stations.tsx]: we don't want to be leaving commented code in when we put up a PR

**Comment by emma-bragg** [comment]
Could we add screenshots to any PRs for frontend changes, e.g. of the drop down here

Also if we could give the PR a more descriptive name than Part1 would be great.

**Comment by emma-bragg** [comment]
With commit messages we want them to be clear what is happening but also not too wordy, think you have some in both directions here :)

File: src/App.css [hunk 2]
```diff
@@ -36,3 +36,26 @@
...
+  margin: 10px;
+}
+
+.dropdown-selections > div {
+  margin: 10px;
+}
+
+.button {
+  height: auto;
+}
```
Comment (emma-bragg) [inline src/App.css]: we want to be adding a newline at the end of all files, as it is a styling convention, github gives this angry red circle when you forget so it's quick to spot when you're looking at your changes

File: src/components/Station.tsx [hunk 3]
```diff
@@ -1,12 +1,15 @@
...
 
-const Station: React.FC = () => {
-    const { id } = useParams();
+export interface StationProps {
+    id: string;
+    name: string;
+}
+
+const Station: React.FC<StationProps> = (attributes) => {
+    // const { id, name } = useParams();
```
Comment (emma-bragg) [inline src/components/Station.tsx]: also here we want to remove commented code, it's good to give a PR  a sanity check of the changes to make sure any comments or console logs etc are removed

File: src/components/Station.tsx [hunk 4]
```diff
@@ -1,12 +1,15 @@
...
 import { useParams } from 'react-router-dom';
 
-const Station: React.FC = () => {
-    const { id } = useParams();
+export interface StationProps {
+    id: string;
+    name: string;
+}
+
+const Station: React.FC<StationProps> = (attributes) => {
```
Comment (emma-bragg) [inline src/components/Station.tsx]: I would use destructing here so you directly reference id and name
```suggestion
const Station: React.FC<StationProps> = ({ id, name }) => {
```

File: src/components/Stations.tsx [hunk 5]
```diff
@@ -1,22 +1,73 @@
...
+    const getCurrentDate = () => {
+        const datetimeString = (new Date()).toISOString();
+        return datetimeString.substring(0, datetimeString.length-1);
+    };
 
-    const [allStations, setAllStations] = useState([]);
+    const handleOnClick = async() => {
+        const baseUrl = 'https://mobile-api-softwire2.lner.co.uk/v1/fares?';
+        const trailingUrl = '&noChanges=false&numberOfAdults=1&numberOfChildren=0&journeyType=single&outboundDateTime=' + getCurrentDate() + '&outboundIsArriveBy=false';
+        const url = baseUrl + 'originStation=' + departureStationChoice + '&destinationStation=' + arrivalStationChoice + trailingUrl;
```
Comment (emma-bragg) [inline src/components/Stations.tsx]: we want all this code doing the api call itself to be in the `ApiCallHelper` file

Comment (emma-bragg) [reply src/components/Stations.tsx]: we can specify these a bit neater using `URLSearchParams` like

```
    const params = new URLSearchParams({
        originStation: departureStationChoice,
        destinationStation: arrivalStationChoice,
        numberOfChildren: '0',
        numberOfAdults: '2',
        ....
    });
```

Comment (emma-bragg) [reply src/components/Stations.tsx]: and then use this like \`${baseUrl}fares?${params.toString()}\` (currently your baseUrl contains the fares text, though other TRAM endpoints might not have it.

In typescript using \`${variable}text\` ([template literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals)) to print you text is best practice instead of using +'s

File: src/components/Stations.tsx [hunk 6]
```diff
@@ -1,22 +1,73 @@
...
+                </div>
+                <div>
+                    <label htmlFor = "arrival stations"> Select the arrival station: </label>
+
+                    <select value = { arrivalStationChoice } name = "stations" id = "arrival-stations" onChange = { (e) => setArrivalStationChoices(e.target.value) }> 
+                        { stationList.map( (stationObj: {id: string; name: string}) => (<Station key = { stationObj.id } id = { stationObj.id } name = { stationObj.name }/>)) }
+                    </select>
+                </div>
+            </div>
+            <button className = "button" type = "button" id = "submitStation" onClick = { handleOnClick } >Submit</button>
```
Comment (emma-bragg) [inline src/components/Stations.tsx]: I don't think we need an id field on the button here.

Similarly on the select I don't think we need name.

Comment (emma-bragg) [inline src/components/Stations.tsx]: alot of this section I'd split onto multiple lines to make it clearer

File: src/components/Stations.tsx [hunk 7]
```diff
@@ -1,22 +1,73 @@
...
     return (
-        <div>
-            Stations!
+        <div className = "station-wrapper">
+            <div className = "dropdown-selections">
+                <div>
+                    <label htmlFor = "departure stations">Select the departure station: </label>
+
+                    <select value = { departureStationChoice } name = "stations" id = "departure-stations" onChange = { (e) => setDepartureStationChoices(e.target.value) }> 
+                        { stationList.map( (stationObj: {id: string; name: string}, index: number) => (<Station key = { index } id = { stationObj.id } name = { stationObj.name }/>)) }
```
Comment (emma-bragg) [inline src/components/Stations.tsx]: we don't need to worry about adding types on to the values we are mapping here as by the definition of map the type of index is defined and we have already clearly defined `stationList `so it knows the type of that already

File: src/components/Stations.tsx [hunk 8]
```diff
@@ -1,22 +1,73 @@
...
+                    </select>
+                </div>
+                <div>
+                    <label htmlFor = "arrival stations"> Select the arrival station: </label>
+
+                    <select value = { arrivalStationChoice } name = "stations" id = "arrival-stations" onChange = { (e) => setArrivalStationChoices(e.target.value) }> 
+                        { stationList.map( (stationObj: {id: string; name: string}) => (<Station key = { stationObj.id } id = { stationObj.id } name = { stationObj.name }/>)) }
+                    </select>
+                </div>
```
Comment (emma-bragg) [inline src/components/Stations.tsx:60]: We could make this nicer by turning all this into a dropdown react component which we could reuse for arrival and departure.

Saying that maybe have a look at using [react-select](https://www.npmjs.com/package/react-select), which is a ready made really nice react dropdown component you run `npm install react-select` and then you have a nice styled dropdown component you can use.

File: src/components/Stations.tsx [hunk 9]
```diff
@@ -1,22 +1,73 @@
...
-    }, []);
+        fetch(url, {
+            headers: {
+                'X-API-KEY': `${process.env.REACT_APP_X_API_KEY}`,
+            }, 
+        })
+            .then(response => response.json())
+            .then(data => data.numberOfChildren)
+            .then(value => alert(value))
+            .then(() => alert(url))
```
Comment (emma-bragg) [inline src/components/Stations.tsx]: I am not sure we want to be alerting here but I am assuming this is a WIP?

File: src/App.tsx [hunk 10]
```diff
@@ -9,7 +9,7 @@ const App = () => (
         <div className = "App">
             <Routes>
                 <Route path = "/stations">
-                    <Route path = ":id" element = { <Station/> }/>
+                    <Route path = ":id" element = { <Station id = 'Paddington' name = 'Paddington'/> }/>
```
Comment (emma-bragg) [inline src/App.tsx]: I'm unsure what you are doing here?

File: src/components/Stations.tsx [hunk 11]
```diff
@@ -1,22 +1,73 @@
...
+    const stationList = [
+        { id: 'PAD', name: 'Paddington' },
+        { id: 'LST', name: 'Liverpool Street' },
+        { id: 'KGX', name: 'Kings Cross' },
+        { id: 'KTN', name: 'Kentish Town' },
+        { id: 'LBG', name: 'London Bridge' },
+    ];
+
+    // let stationChoice: string;
+    const [departureStationChoice, setDepartureStationChoices] = useState('PAD');
```
Comment (emma-bragg) [inline src/components/Stations.tsx]: maybe define 

const initialStation = stationList[0].id

so we wouldn't have to change the text 'PAD' in alot of places if we want to change Paddington to something else

**Comment by emma-bragg** [comment]
Maybe just a me thing as I haven't tested them all but I couldn't find routes between your stations? Which ones do have them?

**Comment by ffiorani** [comment]
refactor: here we are extracting the drop down list into a separate component for reusability. These changes address the comments made above and make the code (hopefully!) look cleaner.

image
 


image
